resources/assets/js/annotations/ol/PolygonValidator.js

```diff
@@ -37,22 +42,25 @@ class PolygonValidator {
             MultiPolygon
         );
 
-        // translate ol geometry into jsts geometry
+        // Translate ol geometry into jsts geometry
         let jstsGeom = parser.read(this.polygon.getGeometry());
-        // divide possibly intersecting polygon at cross points 
-        // in several smaller polygons
-        let possiblyMultiGeom = parser.write(this.jstsValidate(jstsGeom));
-
-        if (possiblyMultiGeom.getCoordinates().length > 1) {
-            // select biggest part
-            let biggestPolygonCoordinates = this.getGreatestPolygonCoordinates(possiblyMultiGeom);
-            this.polygon.getGeometry().setCoordinates(biggestPolygonCoordinates);
+
+        if (jstsGeom.isSimple()) {
             return this.polygon;
         }
 
-        return this.polygon;
+        // Divide non-simple polygon at cross points into several smaller polygons
+        // and translate back to ol geometry
+        let polygons = this.jstsValidate(jstsGeom)['array'].map(p => parser.write(p));
 
+        if (polygons.length > 1) {
+            // Select biggest part
+            let greatestPolygon = this.getGreatestPolygon(polygons);
+            // Only change coordinates because object references are in use
+            this.polygon.getGeometry().setCoordinates(greatestPolygon.getCoordinates());
```

Comment:
This modifies the original polygon, which may be unexpected. Later, you use `setCoordinates()` again (in the annotation canvas/video screen). The validation could be implemented in a functional style instead of a class-based style. So you just implement and export the `isInvalidPolygon(feature)` and `makeValidPolygon(feature)` functions here (without putting them into a class and with all the auxiliary functions they need in this file but not exported). Then `makeValidPolygon()` can directly modify the polygon that was passed as argument.